Translation-equivariant denoisers

jscanvic opened the pull request:
Add `AliasFreeDenoiser`, a translation-equivariant UNet using ideal up and downsampling operators (Karras et al., 2021) and filtered polynomial nonlinearities (Michaeli et al., 2023).

* verify its shift- and translation-equivariance
* verify that UNets are not shift- nor translation-equivariant 
* add the `Translate` class
* add the method `Transform.equivariance_test`

TODO

* [x] add tests for the equivariance of different forward operators
* [x] add an experiment showing how it performs
* [x] add an option for making the network rotation-equivariant as well
* [x] add support for input images with an arbitrary aspect ratio
* [x] add activation normalization

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

Andrewwango commented:
Happy to help review when you're ready @jscanvic

jscanvic commented:
Thanks @Andrewwango! It should be ready for review next week if I manage to get the rotation-equivariance working.

jscanvic commented:
Can you give me some feedback? 🙏 @Andrewwango

Andrewwango commented on deepinv/models/aliasfree.py:
```diff
@@ -0,0 +1,740 @@
+# https://github.com/hmichaeli/alias_free_convnets
```
Could you add a note saying what code is directly from here, what code is adapted from here, and what code is our original code?

Andrewwango replied:
esp with the classes with no documentation - if they're not our implementations that's ok but if they are they should be properly documented

Andrewwango commented on deepinv/models/aliasfree.py:
```diff
@@ -0,0 +1,740 @@
+    return PadLayer
+
+
+class AliasFreeDenoiser(nn.Module):
```
Should this be called AliasFreeUNet? I'm not sure...

jscanvic replied:
Good idea! Another option might be EquivariantUNet. What do you think? I'm changing it to your suggestion for now.

jscanvic replied:
In hindsight I'd say we care more about the equivariance and removing aliasing is just one of the things we need to do to get it. I've renamed it to EquivariantUNet.

Andrewwango reviewed (changes requested):
Thanks for the contributions!

Please can you also add Translate into the test_transforms?

Also unrelated but this might interest you https://docs.e3nn.org/en/latest/guide/equivar_testing.html

Andrewwango commented:
Also changelog please

jscanvic commented on examples/aliasfree/demo_alias_free.py:
```diff
@@ -0,0 +1,138 @@
+        device=device,
+    )
+
+torch.manual_seed(0)
```
The whole "if not exists" is conditional and stochastic: the 1st seed setting makes it deterministic, the 2nd makes what's after deterministic. This is a bit ugly indeed. Two other options would be 1) to fork the rng state in the conditional statement, and 2) to get rid of the conditional and always regenerate the dataset. Any preference?

Andrewwango replied:
Oh I see, indeed that's ugly, and I've never even thought about that before. Note I've just added a conditional overwrite param step in `generate_dataset` too: https://github.com/deepinv/deepinv/blob/859392c97c21f4a14e2fa897a88f4b8602f92c56/deepinv/datasets/datagenerator.py#L108
I think for now, maybe you could remove the first set of random seeding, and just fix the BernoulliSplittingMaskGenerator using a torch.Generator, and keep the second set of random seeding?

Also note that you're fixing the inpainting mask, and not passing the `physics_generator` to `generate_dataset`, so there is no stochasticity anyway in your `if not exists` block?

And you could also use `overwrite_existing` too if it serves you!

jscanvic commented on examples/aliasfree/demo_alias_free.py:
```diff
@@ -0,0 +1,138 @@
+
+from os.path import exists
+
+torch.manual_seed(0)
```
I don't have a strong opinion about that but here are some thoughts.

There should be a way to make everything deterministic in deepinverse, ideally in a straightforward, standard and properly documented manner.

I'm not sure if passing around instances of `torch.Generator` is the perfect solution: how do you deal with the rng from numpy? from the stdlib?

I like the idea of acknowledging some rngs as officially supported by deepinverse and relying on their global state but I reckon it might be problematic for parallel computing.

I'd favor forking rng states over re-seeding rngs as it minimizes interference with end-users' code (it makes the entropy higher too).

I like the idea of having an internal module for messing with the rng but I'm not sure I would expose it to end-users, it might not be the best for interoperability with other libraries.

Hope that helps!

jscanvic commented on deepinv/transform/translate.py:
```diff
@@ -0,0 +1,48 @@
+
+    :param int n_trans: number of transformed versions generated per input image.
+    :param torch.Generator rng: random number generator, if None, use torch.Generator(), defaults to None
+    :param bool constant_shape: ignored, kept for compatibility with other transforms, defaults to True
```
Fixed in e1c128a. Good catch!

jscanvic commented on deepinv/transform/rotate.py:
```diff
@@ -77,3 +91,11 @@ def _transform(
+        if crop_offsets is not None:
+            h1, h2, w1, w2 = crop_offsets
+            x = x[..., h1:h2, w1:w2]
+        elif self.padding is not None:
```
The implementation I'm using for circularly-padded rotations 1) pads the input image, 2) do the zero-padding rotation, and 3) crop the result to the input shape. Step 2) runs for zero-padded and circularly-padded rotations, but step 1) and 3) only run for circularly-padded rotations and not for the zero-padded ones. I'm not exactly sure how to improve that. Any suggestion?

jscanvic commented:
Yesterday I noticed that the implementation only supported square-shaped inputs (mostly due to the implementation of the FFT-based low-pass filters). I've fixed that and added the regression test `test_nonsquare_input`.
